### PR TITLE
Added validation set

### PR DESCRIPTION
**Pull request by Qwer1234fgdgeg**

### CODE REVIEW

File: model/R3D.py [hunk 1]
```diff
@@ -36,8 +36,9 @@ def getSaveDir(self):
             local_time_dir = str(local_time[0]) + '.' + str(local_time[1]) \
                              + '.' + str(local_time[2])
             save_model_path = os.path.join(models_parameters_path, local_time_dir)
-            if not os.path.exists(save_model_path):  # 如果没有这个目录就创建
-                os.mkdir(save_model_path)
+            print(save_model_path)
+            # if not os.path.exists(save_model_path):  # 如果没有这个目录就创建
```
Comment (LHTcode) [inline model/R3D.py]: 这里为什么要注释阿，还有把 print 去掉

File: tools/train.py [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
+sys.path.append(base)
 from data import data_process
 from trainer import common_trainer
 from model.R3D import R3D_18
 from torch.utils.tensorboard import SummaryWriter
 import time
 
-import sys
 
 print(sys.modules)
```
Comment (LHTcode) [inline tools/train.py]: 把print去掉吧

File: tools/train.py [hunk 3]
```diff
@@ -17,7 +19,7 @@ def main():
     root_path = os.path.abspath("../")
 
     # 制作模型
-    train_model = R3D_18(root_path, pretrained=True)  # 使用训练好的参数作为初始化参数
+    train_model = R3D_18('root_path', pretrained=False)  # 使用训练好的参数作为初始化参数
```
Comment (LHTcode) [inline tools/train.py]: 为啥要加引号阿，加引号不就变成了字符串

File: tools/train.py [hunk 4]
```diff
@@ -1,13 +1,15 @@
 import os
-
+import sys
 from torch.utils.data import DataLoader
+
+base = os.path.abspath("../")
+sys.path.append(base)
```
Comment (LHTcode) [inline tools/train.py:5]: 不要写在 import 的中间，而且跟下面的 root_path 重复了，简化一下

File: tools/valid.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import os
+import sys
+from torch.utils.data import DataLoader
+
+base = os.path.abspath("../")
```
Comment (LHTcode) [inline tools/valid.py:5]: 不要写在 import 的中间

File: tools/valid.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+sys.path.append(base)
```
Comment (LHTcode) [inline tools/valid.py:6]: 这两行不要写在 import 的中间

Comment (LHTcode) [inline tools/valid.py:6]: 这两行不要写在 import 中间

Comment (LHTcode) [inline tools/valid.py:5]: 这两行不要写在 import 中间，而且跟下面的 root_path 重复了，简化一下

File: tools/valid.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import time
+import torch
+
+
+#接收模型和数据，进行测试
+def valid_model (model,Dataloader):
+    print("Validating...")
+    #模型的路径
+    root_path = os.path.abspath("../")
+    model_path = root_path+"./models_parameters/2022.7.16/best_model.pt"
```
Comment (LHTcode) [inline tools/valid.py]: 去掉绝对地址，不要出现斜杠

File: tools/valid.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    model.eval()
+    device = model.device
+    model_dict = torch.load(model_path, map_location=torch.device(device))
+    model.load_state_dict(model_dict["state_dict"])
+    print("模型加载完成")
```
Comment (LHTcode) [inline tools/valid.py]: 不要 print 中文信息

File: tools/valid.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+    #进行验证
+    data_num = 0
+    correct_num = 0
+    for data in Dataloader:
+        #视频对应的原标签
+        video, target = data
+        print("target",target)
+        data_num += video.size(dim=0)
+        target = torch.tensor(target).to(torch.long).to(device)
```
Comment (LHTcode) [inline tools/valid.py]: data 返回的 target 就是 tensor，并且不用 .to() 一大堆东西。

File: tools/valid.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+        with torch.set_grad_enabled(False):
+            # 计算输出
+            output = model(video)
+        _, preds = torch.max(output, dim=1)
+        print(output)
+        print("预测",preds)
```
Comment (LHTcode) [inline tools/valid.py]: 不要 print 一些调试信息，并且不要出现中文，要 print 的最关键信息是 precision

File: tools/valid.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        'batch_size': 8,
+        'eval_interval': 1  # 验证间隔，如果设置为1则代表每一次训练的epoch后面都验证一次
+    }
+    valid_dataset = data_process.VideoDataset(root_path, T_size=valider_hyperparam['T_size'], phase="train",
+                                              transform=None, **dataset_setting)
+    Dataloader = DataLoader(valid_dataset, valider_hyperparam['batch_size'], shuffle=True)
+    print("迭代",len(Dataloader))
+
+    root_path = os.path.abspath("../")
+    print(root_path)
```
Comment (LHTcode) [inline tools/valid.py]: 不要 print 调试信息。
**另外，解决完所有的问题之后在本地跑一边，ok了再 commit 上来。**

File: tools/valid.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
+import os
+import sys
+from torch.utils.data import DataLoader
+
+base = os.path.abspath("../")
+sys.path.append(base)
+from data import data_process
+from trainer import common_trainer
+from model.R3D import R3D_18
+from torch.utils.tensorboard import SummaryWriter
```
Comment (LHTcode) [inline tools/valid.py:11]: 你怎么没用到 SummaryWriter ？我要记录精度信息的呀，你把它写进 valid 函数的参数里面，参考一下我的 train 是咋写的。

**Comment by LHTcode** [comment]
还有很多东西没解决。

**Comment by LHTcode** [comment]
我上面那个review那里你还有很多东西没修改的呀，绝对地址、Summarywriter那些。

**Comment by LHTcode** [comment]
你最新的commit就只是加了几个空格和回车 @Qwer1234fgdgeg

**Comment by LHTcode** [comment]
你可以下载个手机邮箱，可以直接在邮件里面回复。**另外**，review 指的是这个：
review截图

**Comment by Qwer1234fgdgeg** [comment]
> 

SummaryWriter有什么要改的

**Comment by LHTcode** [comment]
> > 
> 
> SummaryWriter有什么要改的

你的 valid 里面没写SummaryWirter吧，你要添加一个。

**Comment by LHTcode** [comment]
> > > 
> > 
> > 
> > SummaryWriter有什么要改的
> 
> 你的 valid 里面没写SummaryWirter吧，你要添加一个。

你如果不知道 SummaryWriter 是啥你就先去查查，然后看看我的 train 里面是怎么写的，参考一下。

File: tools/valid.py [hunk 13]
```diff
@@ -43,6 +43,12 @@ def valid_model (model,Dataloader):
         correct_num += torch.sum(preds == target)
     precision = correct_num / data_num * 100  # 计算一个epoch下来的精度
     print("Precision = {}%".format(precision))
+    local_time = time.localtime()
+    local_time_dir = str(local_time[0]) + '.' + str(local_time[1]) \
+                     + '.' + str(local_time[2]) + '.' + str(local_time[3]) \
+                     + '.' + str(local_time[4]) + '.' + str(local_time[5])
+    writer = SummaryWriter(log_dir=os.path.join(root_path, "exact_log", local_time_dir))
```
Comment (LHTcode) [inline tools/valid.py]: 错了错了，这是 wirter 的初始化而已，writer的初始化不应该放在这。并且，log_dir 用我train里面写的那个地址。

**Comment by Qwer1234fgdgeg** [comment]
> 我上面那个review那里你还有很多东西没修改的呀，绝对地址、Summarywriter那些。

**Comment by Qwer1234fgdgeg** [comment]
我全部review看了一次了，还有什么漏了吗

**Comment by LHTcode** [comment]
你怎么把pr关了hhh



---- Replied Message ----
| From | ***@***.***> |
| Date | 07/20/2022 22:20 |
| To | ***@***.***> |
| Cc | ***@***.***>、State ***@***.***> |
| Subject | Re: [LHTcode/Violence-detection] Added validation set (PR #1) |

Closed #1.

**Comment by LHTcode** [comment]
那个summarywriter写错了，你看看我的review那里新的评论



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2022年07月20日 22:21 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>、State ***@***.***> |
| 主题 | Re: [LHTcode/Violence-detection] Added validation set (PR #1) |

我全部review看了一次了，还有什么漏了吗

**Comment by Qwer1234fgdgeg** [comment]
我尝试了一下，结果不会开了

**Comment by Qwer1234fgdgeg** [comment]
我怎么找不到valid的review

**Review by LHTcode — CHANGES_REQUESTED** [review]

**Comment by Qwer1234fgdgeg** [comment]
记录精度要加上epoch吗

**Comment by LHTcode** [comment]
> 记录精度要加上epoch吗

要的

**Comment by LHTcode** [comment]
@Qwer1234fgdgeg 快改

**Comment by Qwer1234fgdgeg** [comment]
好好好，回去马上改

**Comment by LHTcode** [comment]
看了一下，大概ok了

**Review by LHTcode — APPROVED** [review]
